src/box/memtx_sort_data.c

```diff
@@ -0,0 +1,720 @@
+struct memtx_sort_data_reader_entry {
+	/** The entry identifier. */
+	struct memtx_sort_data_key key;
+	/** The offset of the sort data in the file. */
+	long offset;
+	/** The physical size of the sort data. */
+	long psize;
+	/** The amount of tuples stored. */
+	long len;
+};
+
+/** The sort data file header entry (writer version). */
+struct memtx_sort_data_writer_entry {
+	/** The entry identifier. */
+	struct memtx_sort_data_key key;
+	/** Offset of the corresponding entry in the file header. */
+	long header_entry_offset;
+	/** The offset of the sort data in the file. */
+	long offset;
+	/** The physical size of the sort data. */
+	long psize;
+	/** The amount of tuples stored. */
+	long len;
+};
+
+#define mh_name _memtx_sort_data_entries
+#define mh_key_t struct memtx_sort_data_key
+#define mh_node_t struct memtx_sort_data_reader_entry
+#define mh_arg_t void *
+#define mh_hash(a, _) (memtx_sort_data_key_hash((a)->key))
+#define mh_hash_key(a, _) (memtx_sort_data_key_hash(a))
+#define mh_cmp(a, b, _) (memtx_sort_data_key_cmp((a)->key, (b)->key))
+#define mh_cmp_key(a, b, _) (memtx_sort_data_key_cmp(a, (b)->key))
+#define MH_SOURCE
+#include "salad/mhash.h"
+
+#define mh_name _memtx_sort_data_writer_entries
+#define mh_key_t struct memtx_sort_data_key
+#define mh_node_t struct memtx_sort_data_writer_entry
+#define mh_arg_t void *
+#define mh_hash(a, _) (memtx_sort_data_key_hash((a)->key))
+#define mh_hash_key(a, _) (memtx_sort_data_key_hash(a))
+#define mh_cmp(a, b, _) (memtx_sort_data_key_cmp((a)->key, (b)->key))
+#define mh_cmp_key(a, b, _) (memtx_sort_data_key_cmp(a, (b)->key))
+#define MH_SOURCE
+#include "salad/mhash.h"
+
+/** The sort data file writer context. */
+struct memtx_sort_data_writer {
+	/** The sort data file pointer. */
+	FILE *fp;
+	/** The sort data file name. */
+	char filename[PATH_MAX];
+	/** The information about the sort data entries. */
+	struct mh_memtx_sort_data_writer_entries_t *entries;
+	/** The currently handled sort data entry. */
+	struct memtx_sort_data_writer_entry *curr_entry;
+	/** The offset to the next index entry in the file header. */
+	long next_header_entry_offset;
+};
+
+/** The sort data reader context. */
+struct memtx_sort_data_reader {
+	/** The sort data file pointer. */
+	FILE *fp;
+	/** The sort data file name. */
+	char filename[PATH_MAX];
+	/** The information about the sort data entries. */
+	struct mh_memtx_sort_data_entries_t *entries;
+	/** The currently handled sort data entry, NULL if space is skipped. */
+	struct memtx_sort_data_reader_entry *curr_entry;
+	/** The old to new tuple address map used on recovery of a space. */
+	struct mh_ptrptr_t *old2new;
+	/** The buffer to use on fread. */
+	void *buffer;
+};
+
+/* {{{ Utilities **************************************************************/
+
+/** Sort data file header format strings. */
+static const char *ENTRIES_FMT = "Entries: %010u\n";
+static const char *ENTRY_FMT = "%010u/%010u: %016lx, %016lx, %20li\n";
```

Comment:
Panic on corrupted file is bad. Let's calculate `crc` checksum for each index sortdata and write it to the file.
The checksum should be added as a new attribute to `ENTRY_FMT`.

Also, we should add a new attribute meaning checksum type to make it extensible in the future.

And, it would be nice to perform a small research on `crc32` efficiency on large amounts of data (in terms of collisions).